File: pypp_cli/test_dir/cpp/src/number_types/in_collections.cpp
#include "number_types/in_collections.h"
#include "py_dict.h"
#include "py_list.h"
#include "py_set.h"
#include "py_str.h"
#include "pypp_util/print.h"
#include <cstdint>

void _lists() {
    pypp::PyList<int8_t> a({1, 2, 3, 4});
    pypp::print(a);
    pypp::PyList<int16_t> b({5, 6, 7, 8});
    pypp::print(b);
    pypp::PyList<int32_t> c({9, 10, 11, 12});
    pypp::print(c);
    pypp::PyList<int64_t> d({13, 14, 15, 16});
    pypp::print(d);
    pypp::PyList<uint8_t> e({17, 18, 19, 20});
    pypp::print(e);
    pypp::PyList<uint16_t> f({21, 22, 23, 24});
    pypp::print(f);
    pypp::PyList<uint32_t> g({25, 26, 27, 28});
    pypp::print(g);
    pypp::PyList<uint64_t> h({29, 30, 31, 32});
    pypp::print(h);
    pypp::PyList<float> i({3.14, 2.71, 1.61});
    pypp::print(i);
}

void _sets() {
    pypp::PySet<int8_t> a({1, 2, 3, 4});
    pypp::print(a);
    pypp::PySet<int16_t> b({5, 6, 7, 8});
    pypp::print(b);
    pypp::PySet<int32_t> c({9, 10, 11, 12});
    pypp::print(c);
    pypp::PySet<int64_t> d({13, 14, 15, 16});
    pypp::print(d);
    pypp::PySet<uint8_t> e({17, 18, 19, 20});
    pypp::print(e);
    pypp::PySet<uint16_t> f({21, 22, 23, 24});
    pypp::print(f);
    pypp::PySet<uint32_t> g({25, 26, 27, 28});
    pypp::print(g);
    pypp::PySet<uint64_t> h({29, 30, 31, 32});
    pypp::print(h);
    pypp::PySet<float> i({3.14, 2.71, 1.61});
    pypp::print(i);
}

void _dicts() {
    pypp::PyDict<int8_t, pypp::PyStr> a = {{1, pypp::PyStr("a")},
                                           {2, pypp::PyStr("b")}};
    pypp::print(a);
    pypp::PyDict<int16_t, pypp::PyStr> b = {{3, pypp::PyStr("c")},
                                            {4, pypp::PyStr("d")}};
    pypp::print(b);
    pypp::PyDict<int32_t, pypp::PyStr> c = {{5, pypp::PyStr("e")},
                                            {6, pypp::PyStr("f")}};
    pypp::print(c);
    pypp::PyDict<int64_t, pypp::PyStr> d = {{7, pypp::PyStr("g")},
                                            {8, pypp::PyStr("h")}};
    pypp::print(d);
    pypp::PyDict<uint8_t, pypp::PyStr> e = {{9, pypp::PyStr("i")},
                                            {10, pypp::PyStr("j")}};
    pypp::print(e);
    pypp::PyDict<uint16_t, pypp::PyStr> f = {{11, pypp::PyStr("k")},
                                             {12, pypp::PyStr("l")}};
    pypp::print(f);
    pypp::PyDict<uint32_t, pypp::PyStr> g = {{13, pypp::PyStr("m")},
                                             {14, pypp::PyStr("n")}};
    pypp::print(g);
    pypp::PyDict<uint64_t, pypp::PyStr> h = {{15, pypp::PyStr("o")},
                                             {16, pypp::PyStr("p")}};
    pypp::print(h);
    pypp::PyDict<float, pypp::PyStr> i = {{3.14, pypp::PyStr("q")},
                                          {2.71, pypp::PyStr("r")}};
    pypp::print(i);
    pypp::PyDict<int16_t, float> j = {{1, 3.14}, {2, 2.71}};
    pypp::print(j);
    pypp::PyDict<uint16_t, int32_t> k = {{1, 1}, {2, 2}};
    pypp::print(k);
}

pypp::PyList<uint16_t> _list_factory() {
    pypp::PyList<uint16_t> ret({1, 2, 3, 4});
    return ret;
}

pypp::PySet<uint16_t> _set_factory() {
    pypp::PySet<uint16_t> ret({1, 2, 3, 4});
    return ret;
}

pypp::PyDict<uint16_t, pypp::PyStr> _dict_factory() {
    return {{1, pypp::PyStr("a")}, {2, pypp::PyStr("b")}};
}

void number_types_in_collections_fn() {
    pypp::print(pypp::PyStr("NUMBER TYPES IN COLLECTIONS RESULTS:"));
    _lists();
    _sets();
    _dicts();
    pypp::print(_list_factory());
    pypp::print(_set_factory());
    pypp::print(_dict_factory());
}

File: test_dir/src/number_types/in_collections.py
from pypp_python import (
    int8_t,
    int32_t,
    int64_t,
    int16_t,
    uint8_t,
    uint16_t,
    uint32_t,
    uint64_t,
    float32,
)


def _lists():
    a: list[int8_t] = [1, 2, 3, 4]
    print(a)
    b: list[int16_t] = [5, 6, 7, 8]
    print(b)
    c: list[int32_t] = [9, 10, 11, 12]
    print(c)
    d: list[int64_t] = [13, 14, 15, 16]
    print(d)
    e: list[uint8_t] = [17, 18, 19, 20]
    print(e)
    f: list[uint16_t] = [21, 22, 23, 24]
    print(f)
    g: list[uint32_t] = [25, 26, 27, 28]
    print(g)
    h: list[uint64_t] = [29, 30, 31, 32]
    print(h)
    i: list[float32] = [3.14, 2.71, 1.61]
    print(i)


def _sets():
    a: set[int8_t] = {1, 2, 3, 4}
    print(a)
    b: set[int16_t] = {5, 6, 7, 8}
    print(b)
    c: set[int32_t] = {9, 10, 11, 12}
    print(c)
    d: set[int64_t] = {13, 14, 15, 16}
    print(d)
    e: set[uint8_t] = {17, 18, 19, 20}
    print(e)
    f: set[uint16_t] = {21, 22, 23, 24}
    print(f)
    g: set[uint32_t] = {25, 26, 27, 28}
    print(g)
    h: set[uint64_t] = {29, 30, 31, 32}
    print(h)
    i: set[float32] = {3.14, 2.71, 1.61}
    print(i)


def _dicts():
    a: dict[int8_t, str] = {1: "a", 2: "b"}
    print(a)
    b: dict[int16_t, str] = {3: "c", 4: "d"}
    print(b)
    c: dict[int32_t, str] = {5: "e", 6: "f"}
    print(c)
    d: dict[int64_t, str] = {7: "g", 8: "h"}
    print(d)
    e: dict[uint8_t, str] = {9: "i", 10: "j"}
    print(e)
    f: dict[uint16_t, str] = {11: "k", 12: "l"}
    print(f)
    g: dict[uint32_t, str] = {13: "m", 14: "n"}
    print(g)
    h: dict[uint64_t, str] = {15: "o", 16: "p"}
    print(h)
    i: dict[float32, str] = {3.14: "q", 2.71: "r"}
    print(i)
    j: dict[int16_t, float32] = {1: 3.14, 2: 2.71}
    print(j)
    k: dict[uint16_t, int32_t] = {1: 1, 2: 2}
    print(k)


def _list_factory() -> list[uint16_t]:
    # Note: you have to explicitly create the variable and return it.
    #  You can't just do: return [1, 2, 3, 4]. Because the type can't be inferred
    #  properly by the C++ compiler.
    ret: list[uint16_t] = [1, 2, 3, 4]
    return ret


def _set_factory() -> set[uint16_t]:
    ret: set[uint16_t] = {1, 2, 3, 4}
    return ret


def _dict_factory() -> dict[uint16_t, str]:
    return {1: "a", 2: "b"}


def number_types_in_collections_fn():
    print("NUMBER TYPES IN COLLECTIONS RESULTS:")
    _lists()
    _sets()
    _dicts()
    print(_list_factory())
    print(_set_factory())
    print(_dict_factory())
